skbase/tests/test_base.py

```diff
@@ -251,26 +484,332 @@ def test_reset_composite():
+    base_obj_set_params = base_obj.set_params()
+    assert (
+        isinstance(base_obj_set_params, FIXTURE_EXAMPLE)
+        and base_obj_set_params.get_params() == orig_params
+    )
+
+
+# This section tests the clone functionality
+# These have been adapted from sklearn's tests of clone to use the clone
+# method that is included as part of the BaseObject interface
+def test_clone():
+    """Test that clone is making a deep copy as expected."""
+    # Creates a BaseObject and makes a copy of its original state
+    # (which, in this case, is the current state of the BaseObject),
+    # and check that the obtained copy is a correct deep copy.
+    base_obj = FIXTURE_EXAMPLE(a=7.0, b="some_str")
+    new_base_obj = base_obj.clone()
+    assert base_obj is not new_base_obj
+    assert base_obj.get_params() == new_base_obj.get_params()
+
+
+def test_clone_2():
+    """Test that clone does not copy attributes not set in constructor."""
+    # We first create an estimator, give it an own attribute, and
+    # make a copy of its original state. Then we check that the copy doesn't
+    # have the specific attribute we manually added to the initial estimator.
+
+    base_obj = FIXTURE_EXAMPLE(a=7.0, b="some_str")
+    base_obj.own_attribute = "test"
+    new_base_obj = base_obj.clone()
+    assert not hasattr(new_base_obj, "own_attribute")
+
+
+def test_clone_raises_error_for_nonconforming_objects():
+    """Test that clone raises an error on nonconforming BaseObjects."""
+    buggy = FIXTURE_BUGGY()
+    buggy.a = 2
+    with pytest.raises(RuntimeError):
+        buggy.clone()
+
+    varg_obj = FIXTURE_INVALID_INIT(a=7)
+    with pytest.raises(RuntimeError):
+        varg_obj.clone()
+
+    obj_that_modifies = FIXTURE_MODIFY_PARAM(a=[0])
+    with pytest.raises(RuntimeError):
+        obj_that_modifies.clone()
+
+
+# def test_clone_empty_array():
+#     # Regression test for cloning estimators with empty arrays
+#     clf = MyEstimator(empty=np.array([]))
+#     clf2 = clone(clf)
+#     assert_array_equal(clf.empty, clf2.empty)
+
+#     clf = MyEstimator(empty=sp.csr_matrix(np.array([[0]])))
+#     clf2 = clone(clf)
+#     assert_array_equal(clf.empty.data, clf2.empty.data)
+
+
+# def test_clone_nan():
+#     # Regression test for cloning estimators with default parameter as np.nan
+#     clf = MyEstimator(empty=np.nan)
+#     clf2 = clone(clf)
+
+#     assert clf.empty is clf2.empty
+
+
+def test_clone_estimator_types():
+    """Test clone works for parameters that are types rather than instances."""
+    base_obj = Example(c=Example)
+    new_base_obj = base_obj.clone()
+
+    assert base_obj.c == new_base_obj.c
+
+
+def test_clone_class_rather_than_instance_raises_error():
+    """Test clone raises expected error when cloning a class instead of an instance."""
+    msg = "You should provide an instance of scikit-learn estimator"
+    with pytest.raises(TypeError, match=msg):
+        clone(FIXTURE_EXAMPLE)
```

Comment:
do we want to have an extraneous clone function?

Reply:
That's a good question. We are currently coupled with scikit-learn's `BaseEstimator` and separate clone function.

So for the time being, I think it makes sense to test the expected error is raised when interacting with the cloned functionality in scikit-learn. Eventually, we might remove the test (and the error raising it is related to) in the skbase `clone` method, because it is not a class method, and therefore, would be invoked on an instance by default.

Reply:
For example, we might decide to have a clone function that clones duck-typed BaseObjects that comply with most of the interface, but don't have their own clone method.

Reply:
ok, makes sense as an interface compatibility test, with `scikit-learn`. That's a separate topic from whether we want to expose a separate `clone` ourselves.